postgres/connection/connection.go

```diff
@@ -15,52 +15,81 @@
 
-// connBuffers maintains a pool of buffers, reusable between connections.
+// connBuffers maintains a pool of buffers, reusable between connections. These are only used for processing
+// fixed-length messages where we know we won't exceed the buffer size on a single read.
 var connBuffers = sync.Pool{
 	New: func() any {
-		return make([]byte, bufferSize)
+		return make([]byte, BufferSize)
 	},
 }
 
-var sliceOfZeroes = make([]byte, bufferSize)
+const headerSize = 5
 
-// Receive returns all messages that were sent from the given connection. A connection may send multiple messages at
-// once, therefore this may return multiple messages. This checks with all messages that have a Header, and have
-// called AddMessageHeader within their init() function. Returns a nil slice if no messages were matched. This is the
-// recommended way to check for messages when a specific message is not expected. Use ReceiveInto or ReceiveIntoAny when
-// expecting specific messages, where it would be an error to receive messages different from the expectation.
-func Receive(conn net.Conn) ([]Message, error) {
-	buffer := connBuffers.Get().([]byte)
-	defer connBuffers.Put(zeroBuffer(buffer))
+// headerBuffers maintains a pool of buffers, reusable between connections, that are used for reading message headers
+// (the first 5 bytes of a client message)
+var headerBuffers = sync.Pool{
+	New: func() any {
+		return make([]byte, headerSize)
+	},
+}
 
-	if _, err := conn.Read(buffer); err != nil {
+var sliceOfZeroes = make([]byte, BufferSize)
+
+// Receive returns all messages that were sent from the given connection. This checks with all messages that have a
+// Header, and have called AddMessageHeader within their init() function. Returns a nil slice if no messages were
+// matched. This is the recommended way to check for messages when a specific message is not expected.
+// Use ReceiveInto or ReceiveIntoAny when expecting specific messages, where it would be an error to receive messages
+// different from the expectation.
+func Receive(conn net.Conn) (Message, error) {
+	header := headerBuffers.Get().([]byte)
+	defer headerBuffers.Put(header)
+
+	n, err := conn.Read(header)
+	if err != nil {
 		return nil, err
 	}
-	db := newDecodeBuffer(buffer)
-	var outMessages []Message
-	for len(db.data) > 0 {
-		message, ok := allMessageHeaders[db.data[0]]
-		if !ok {
-			break
-		}
-		outMessage, err := receiveFromBuffer(db, message)
-		if err != nil {
-			return nil, err
-		}
-		outMessages = append(outMessages, outMessage)
-		db.next()
+
+	if n < headerSize {
+		return nil, errors.New("received message header is too short")
 	}
-	return outMessages, nil
+
+	message, ok := allMessageHeaders[header[0]]
+	if !ok {
+		return nil, fmt.Errorf("received message header is not recognized: %v", header[0])
+	}
+
+	// TODO: possibly not every message has a length in this position, need an easy interface to tell us if so
+	messageLen := int(binary.BigEndian.Uint32(header[1:])) - 4
```

Comment:
This all sounds fine except the for uint32 bit -- there's no binary.BigEndian.Int32, it always deals with unsigned ints. I can add a max check if you want

Reply:
So I added verification in message_verification, and discovered 4 messages with lengths not preceded by headers:

* All 3 startup message types
* CancelRequest

Not sure how to best handle this, going to leave it alone for now.

Reply:
I misspoke earlier, I meant that, if a header is present, then a message length always directly follows. None of the messages you mentioned have a header, so that’s fine.

Reply:
You’re correct about `BigEndian`. I cast `int32` to and from `uint32` during encoding and decoding. So perhaps we should have a max check? Perhaps in encoding as well.

Reply:
That makes more sense lol